compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -0,0 +1,802 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering from the AMDAIE dialect to AIE and AIEX
+// dialects.
+//
+//===----------------------------------------------------------------------===//
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIEX/IR/AIEXDialect.h"
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-amdaie-lower-to-aie"
+
+using namespace xilinx;
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Utility to remap the provided operation's operands.
+void remapOperands(Operation *op, IRMapping &mapper) {
+  for (int i = 0; i < op->getNumOperands(); ++i) {
+    Value operand = op->getOperand(i);
+    if (mapper.contains(operand)) {
+      op->setOperand(i, mapper.lookup(operand));
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Convert amdaie.core operation to aie.core
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Utility to create an `aie.objectfifo` operation from
+/// `amdaie.circular_dma_cpy_nd`.
+AIE::ObjectFifoCreateOp createObjectFifo(IRRewriter &rewriter,
+                                         AMDAIE::CircularDmaCpyNdOp dmaOp,
+                                         Value srcTile, ValueRange dstTiles,
+                                         StringAttr &symName) {
+  // Create BD layout attributes from DMA `strides` and `sizes`.
+  SmallVector<AIE::BDDimLayoutAttr, 4> srcBDDimLayoutAttrs;
+  for (auto [size, stride] :
+       llvm::zip(dmaOp.getSourceMixedSizes(), dmaOp.getSourceMixedStrides())) {
+    srcBDDimLayoutAttrs.push_back(AIE::BDDimLayoutAttr::get(
+        rewriter.getContext(), getConstantIntValue(size).value(),
+        getConstantIntValue(stride).value()));
+  }
+  SmallVector<AIE::BDDimLayoutAttr, 4> dstBDDimLayoutAttrs;
+  for (auto [size, stride] :
+       llvm::zip(dmaOp.getTargetMixedSizes(), dmaOp.getTargetMixedStrides())) {
+    dstBDDimLayoutAttrs.push_back(AIE::BDDimLayoutAttr::get(
+        rewriter.getContext(), getConstantIntValue(size).value(),
+        getConstantIntValue(stride).value()));
+  }
+  AIE::BDDimLayoutArrayAttr srcDims = AIE::BDDimLayoutArrayAttr::get(
+      rewriter.getContext(), ArrayRef(srcBDDimLayoutAttrs));
+  SmallVector<AIE::BDDimLayoutArrayAttr> dstDimsVec;
+  dstDimsVec.push_back(AIE::BDDimLayoutArrayAttr::get(
+      rewriter.getContext(), ArrayRef(dstBDDimLayoutAttrs)));
```

Comment:
Because `AIE::ObjectFifoCreateOp` can handle multiple consumers with a different strided access pattern, but not multiple producers: https://github.com/Xilinx/mlir-aie/blob/66e642b1bf6c520f233c8161bb533e788dd9731d/include/aie/Dialect/AIE/IR/AIEOps.td#L1611